algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -206,64 +195,73 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-                    })
-                    .collect::<Vec<_>>();
-                (circuit_id, z_b_i)
-            })
-            .collect::<BTreeMap<_, _>>();
+        // We're now going to calculate the rowcheck_zerocheck
+        let rowcheck_time = start_timer!(|| "Rowcheck");
+
+        let v_R_at_alpha_time = start_timer!(|| "v_R_at_alpha");
+        let v_R_at_alpha = max_constraint_domain.evaluate_vanishing_polynomial(alpha);
+        end_timer!(v_R_at_alpha_time);
+
+        #[rustfmt::skip]
+        let rowcheck_zerocheck = {
+            let mut rowcheck_zerocheck = LinearCombination::empty("rowcheck_zerocheck");
+            for (i, (id, c)) in batch_combiners.iter().enumerate() {
+                let mut circuit_term = LinearCombination::empty(format!("rowcheck_zerocheck term {id}"));
+                let third_sums_i = &prover_third_message.sums[i];
+                let circuit_state = &state.circuit_specific_states[id];
+
+                for (j, instance_combiner) in c.instance_combiners.iter().enumerate() {
+                    let mut rowcheck = LinearCombination::empty(format!("rowcheck term {id}"));
+                    let sum_a_third = third_sums_i[j].sum_a;
+                    let sum_b_third = third_sums_i[j].sum_b;
+                    let sum_c_third = third_sums_i[j].sum_c;
+
+                    rowcheck.add(sum_a_third*sum_b_third - sum_c_third, LCTerm::One);
+
+                    circuit_term += (*instance_combiner, &rowcheck);
+                }
+                let constraint_domain = circuit_state.constraint_domain;
+                let selector = selector_evals(&mut selectors, &max_constraint_domain, &constraint_domain, alpha);
+                circuit_term *= selector;
+                rowcheck_zerocheck += (c.circuit_combiner, &circuit_term);
+            }
+            rowcheck_zerocheck.add(-v_R_at_alpha, "h_0");
+            rowcheck_zerocheck
+        };
+
+        debug_assert!(evals.get_lc_eval(&rowcheck_zerocheck, alpha)?.is_zero());
+        linear_combinations.insert("rowcheck_zerocheck".into(), rowcheck_zerocheck);
+        end_timer!(rowcheck_time);
+
+        // Lineval sumcheck:
+        let lineval_time = start_timer!(|| "Lineval");
 
         let g_1 = LinearCombination::new("g_1", [(F::one(), "g_1")]);
 
-        let bivariate_poly_time = start_timer!(|| "Bivariate poly");
-        let mut r_alpha_at_beta_s = BTreeMap::new();
-        for (circuit_id, circuit_state) in state.circuit_specific_states.iter() {
-            r_alpha_at_beta_s.insert(
-                circuit_id,
-                circuit_state.constraint_domain.eval_unnormalized_bivariate_lagrange_poly(alpha, beta),
-            );
-        }
-        end_timer!(bivariate_poly_time);
+        let v_C_at_beta_time = start_timer!(|| "v_C_at_beta");
+        let v_C_at_beta = max_variable_domain.evaluate_vanishing_polynomial(beta);
+        end_timer!(v_C_at_beta_time);
```

Comment:
I think we can remove this timer; the evaluation time should be close to zero.
```suggestion
        let v_C_at_beta = max_variable_domain.evaluate_vanishing_polynomial(beta);
```

Reply:
47240504d